rust/azure_iot_operations_connector/src/data_processor/derived_json.rs

```diff
@@ -0,0 +1,704 @@
+        }"#;
+        let input_json: Value = serde_json::from_str(input_json_str).unwrap();
+
+        let expected_output_json_str = r#"{
+            "factory": "home",
+            "temperature": 10,
+            "metadata": null
+        }"#;
+        let expected_output_json: Value = serde_json::from_str(expected_output_json_str).unwrap();
+
+        let dataset = Dataset {
+            dataset_configuration: None,
+            data_points: vec![
+                // This datapoint is unchanged in the transformation
+                DatasetDataPoint {
+                    name: String::from("factory"),
+                    data_source: String::from("factory"),
+                    type_ref: None,
+                    data_point_configuration: None,
+                },
+                // This datapoint is renamed in the same scope
+                DatasetDataPoint {
+                    name: String::from("temperature"),
+                    data_source: String::from("temp"),
+                    type_ref: None,
+                    data_point_configuration: None,
+                },
+                // This datapoint does not exist in the input json
+                DatasetDataPoint {
+                    name: String::from("metadata"),
+                    data_source: String::from("metadata"),
+                    type_ref: None,
+                    data_point_configuration: None,
+                },
+            ],
+            data_source: None,
+            destinations: vec![],
+            name: "TestDataset".to_string(),
+            type_ref: None,
+        };
+
+        // Metadata being null is not enough information to derive type information about the field
+        // and so it is simply inferred as being "true"
+        let expected_output_json_schema_str = r#"{
+            "$schema": "http://json-schema.org/draft-07/schema#",
+            "type": "object",
+            "properties": {
+                "factory": {
+                    "type": "string"
+                },
+                "temperature": {
+                    "type": "integer"
+                },
+                "metadata": true
+        }
+        }"#;
+        let expected_output_json_schema: Value =
+            serde_json::from_str(expected_output_json_schema_str).unwrap();
+
+        TransformTestCase {
+            dataset,
+            input_json,
+            expected_output_json,
+            expected_output_json_schema,
+        }
+    }
+
+    #[test_case(&valid_testcase_1(); "1:1 transformation")]
+    #[test_case(&valid_testcase_2(); "Subset transformation")]
+    #[test_case(&valid_testcase_3(); "Overlapping transformation")]
+    #[test_case(&valid_testcase_4(); "No datapoints")]
+    #[test_case(&valid_testcase_5(); "Missing data source")]
+    fn valid_transform(test_case: &TransformTestCase) {
+        let input_data = Data {
+            payload: serde_json::to_vec(&test_case.input_json).unwrap(),
+            content_type: None,
```

Comment:
fyi content type is going to have to be required since we can't send telemetry without it, so I'd go ahead and have these be Some("application/json") to reduce churn later

Reply:
good catch